src/spclient.rs

```diff
@@ -1,77 +1,135 @@
 use bip39::Mnemonic;
 use bitcoin::{
-    secp256k1::{Secp256k1, SecretKey},
+    secp256k1::{Secp256k1, SecretKey, ONE_KEY},
     util::bip32::{DerivationPath, ExtendedPrivKey},
-    Network,
+    Network, OutPoint, Script, Txid,
 };
-use once_cell::sync::OnceCell;
+use serde::{Serialize, Deserialize};
 use silentpayments::receiving::Receiver;
 use std::str::FromStr;
 
-use lazy_static::lazy_static;
 use anyhow::Result;
 
-lazy_static! {
-    static ref SPCLIENT: OnceCell<SpClient> = OnceCell::new();
-}
-
-pub fn create_sp_client(
-    scan_sk: SecretKey,
-    spend_sk: SecretKey,
-    birthday: u32,
-    is_testnet: bool,
-) -> Result<()> {
-    let spclient = SpClient::new(scan_sk, spend_sk, is_testnet, birthday)?;
+use crate::db::FileWriter;
 
-    let _ = SPCLIENT.set(spclient);
-    Ok(())
+pub struct ScanProgress {
+    pub start: u32,
+    pub current: u32,
+    pub end: u32,
 }
 
-pub fn get_sp_client() -> &'static SpClient {
-    SPCLIENT.wait()
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct OwnedOutput {
+    pub txoutpoint: OutPoint, 
+    pub blockheight: u32,
+    pub tweak: String,
+    pub amount: u64,
+    pub script: Script,
+    pub spent: bool,
+    pub spent_by: Option<Txid>
 }
 
-pub fn get_receiving_address() -> Result<String> {
-    let client = get_sp_client();
-
-    let receiver = &client.sp_receiver;
-
-    Ok(receiver.get_receiving_address())
-}
-
-pub fn get_birthday() -> u32 {
-    let client = get_sp_client();
-    let birthday = client.birthday;
-    birthday
-}
 
-#[derive(Debug)]
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
 pub struct SpClient {
-    pub scan_sk: SecretKey,
-    pub spend_sk: SecretKey,
+    pub label: String,
+    scan_sk: SecretKey,
+    spend_sk: SecretKey,
     pub sp_receiver: Receiver,
     pub birthday: u32,
+    pub last_scan: u32,
+    pub total_amt: u64,
+    owned: Vec<OwnedOutput>,
+    writer: FileWriter,
 }
 
 impl SpClient {
     pub fn new(
+        label: String,
         scan_sk: SecretKey,
         spend_sk: SecretKey,
-        is_testnet: bool,
         birthday: u32,
+        is_testnet: bool,
+        path: String,
     ) -> Result<Self> {
         let secp = Secp256k1::signing_only();
-        let spend_pubkey = spend_sk.public_key(&secp);
         let scan_pubkey = scan_sk.public_key(&secp);
+        let spend_pubkey = spend_sk.public_key(&secp);
         let sp_receiver = Receiver::new(0, scan_pubkey, spend_pubkey, is_testnet)?;
+        let writer = FileWriter::new(path, label.clone())?;
 
         Ok(Self {
+            label,
             scan_sk,
-            spend_sk,
             sp_receiver,
             birthday,
+            last_scan: birthday,
```

Comment:
I think `last_scan` means the latest block that has been scanned, so this should probably be `birthday - 1` to be completely accurate. (Also may need to be changed in reset_wallet)

Reply:
Agree, but gotta watch out for birthday 0 though :)

Ok for reset_from_blockheight